### ISSUE TITLE
`num_boost_round ` and `early_stopping_rounds` in xgboost.train() API

### ISSUE THREAD

**Issue by wyd1582** [dmlc/xgboost#4909]
Hi,

I am confused if we use early stopping with number of boost round larger than 1, are we still getting multiple boosted random forest and each rf has `n_estimators` trees? 

Thanks

**Comment by trivialfis** [comment]
For `xgboost.train`,  I don't think there's an `n_estimators` parameter.  Use `num_parallel_tree` for boosting forest.

**Comment by trivialfis** [comment]
Each boosting rounds adds a random forest with `num_parallel_tree` trees.  Hope that helps.

**Comment by wyd1582** [comment]
cool, however if i use sklearn xgboost API, when i call XGBClassifier.fit(), i dont get to pass 'num_boost_round' anyways which meaning i am only trianing boosting trees within 'n_estimators'? depend on my 'early_stopping_round' set up? thanks in advance

**Comment by trivialfis** [comment]
`n_estimators` is an scikit-learn version of `num_boost_round` I think.

**Comment by wyd1582** [comment]
oh , so assume i pass `n_estimators` 100, i train on 50 iterations with `early_stopping_round` then how many random forests and how many trees i end up get ? thanks

**Comment by trivialfis** [comment]
100 `n_estimators` means 100 iterations, resulting in 100 stacked trees.  If you want to do random forest, with native API that would be:
```
xgboost.train({'num_parallel_tree': 50, 'tree_method`: ... }, X, num_boost_round=100)
```

The above line will result in 100 stacked random forests with 50 trees each.  The sklearn API is mostly a wrapper around above native API.

**Comment by wyd1582** [comment]
i see so with sklearn, you can only do boost trees instead of RFs. Thanks for confirm,rly helpful thanks

**Comment by trivialfis** [comment]
For non-boosted random forest (only 1 random forest), you can use `XGBRFRegressor` and `XGBRFClassifier`.

**Comment by trivialfis** [comment]
You can do boosting RF in sklearn.  Just provide `num_parallel_tree` as part of the `kwargs`.  Or use `set_param` class method.

**Comment by wyd1582** [comment]
sorry could you be more specify how do i train rf booster with `XGBClassifier`  in sklearn xgboost API? `n_estimators` and `num_parallel_tree` both larger than `?

**Comment by trivialfis** [comment]
``` python
import numpy as np
import xgboost

kRows = 1000
kCols = 100

X = np.random.randn(kRows, kCols)
y = np.random.randint(low=0, high=4, size=kRows)

dtrain = xgboost.DMatrix(X, y)

cls = xgboost.XGBClassifier(
    n_estimators=100,
    n_jobs=8,
    # following belongs to kwargs so may not work with some
    # scikit-learn methods
    num_parallel_tree=8,
    tree_method='hist'
)

cls.fit(X, y)
```

Be careful that due to multi-class training uses one tree for each class.  So when you set `num_parallel_tree` to 8 and with 4 classes, you get 32 new trees for each iteration, with 100 iterations you will have total 3200 trees in final booster ...

@hcho3 Correct me if I'm wrong.  Also we need to revisit the sklearn wrapper for updating parameters when time allows, so I'm keeping this issue open. ;-)

**Comment by wyd1582** [comment]
good, will follow this practice and interested to see the result then, rly helpful!

**Comment by hcho3** [comment]
@trivialfis I think your explanation is reasonable.

**Comment by wyd1582** [comment]
so if i pass `n_estimator` (`num_parallel_tree`) to larger than 1 and also `num_boost_round` larger to one, what will happen, what will `n_estimator` convert to and am i also training boosted rfs?

**Comment by hcho3** [comment]
@wyd1582 You should not pass `num_boost_round` since you are using the scikit-learn interface.

**Comment by wyd1582** [comment]
i mean when i was using xgboost.train() the learning API not the `XGBClassifier`- sorry to ask back and force cause these parameters are pretty confusing for me now

**Comment by hcho3** [comment]
@wyd1582 You should pick which interface to use.
* If using `xgboost.train()` : use `num_boost_round`
* If using `XGBClassifier`: use `n_estimator`

**Comment by wyd1582** [comment]
sorry but i actually trained an object with this set up : now i am confused what i trained:
>
XGB_KWARGS= dict(num_boost_round = 50)
PARAM = dict(silent=1, n_estimators = 100, max_depth= 7, gamma =0.01,
               min_child_weight=1, max_delta_step=1, subsample=1, colsample_bytree=1, 
               colsample_bylevel=1,  scale_pos_weight=1, 
               base_score=0.05, seed=0, missing=None, random_state=0,
               obj= 'binary:logistic')
>
 and do 
>
xgb.train(** PARAM, **XGB_KWARGS)
> 
am i training 50 boosted RFs each has 100 trees??

**Comment by hcho3** [comment]
`n_estimators` should not be there, because you are using `xgb.train()`. So `n_estimators` has no effect.

**Comment by wyd1582** [comment]
cool if it is confirmed then i do understand the relation with those parameters now, so with this set up i am training boosted trees no **Random Forests**
thanks 👍

**Comment by wyd1582** [comment]
and to confirm, my `num_parallel_tree` by defualt is 1 , and if i don't pass that set up , i will also only boost one tree each iteration

**Comment by hcho3** [comment]
Yes

**Comment by wyd1582** [comment]
this really solved a lot confusions thanks
